File: HTCommander.Core/Utils/RigctldServer.cs
/*
Copyright 2026 Ylian Saint-Hilaire
Licensed under the Apache License, Version 2.0 (the "License");
http://www.apache.org/licenses/LICENSE-2.0
*/

// Protocol: Hamlib rigctld text protocol
// Reference: https://hamlib.sourceforge.net/manuals/hamlib.html

using System;
using System.Collections.Concurrent;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace HTCommander
{
    /// <summary>
    /// Cross-platform rigctld TCP server for PTT and rig control.
    /// Used by fldigi, WSJT-X, Direwolf, VaraFM (via Hamlib mode).
    /// Follows the AgwpeServer DataBroker integration pattern.
    /// </summary>
    public class RigctldServer : IDisposable
    {
        private DataBrokerClient broker;
        private TcpListener listener;
        private readonly ConcurrentDictionary<Guid, RigctldClientHandler> clients = new ConcurrentDictionary<Guid, RigctldClientHandler>();
        private const int MaxClients = 10;
        private CancellationTokenSource cts;
        private Task serverTask;
        private int port;
        private volatile bool running = false;
        private bool bindAll = false;
        private volatile bool pttActive = false;
        private Timer pttSilenceTimer;
        private Timer pttTimeoutTimer;
        private const int PttTimeoutMs = 30000;
        private readonly object pttLock = new object();
        private long cachedFrequency = 145500000; // Accessed from multiple threads; reads/writes are non-atomic on 32-bit but acceptable for cached display value
        private volatile int activeRadioId = -1;

        public bool PttActive => pttActive;
        public long CachedFrequency => cachedFrequency;

        public RigctldServer()
        {
            broker = new DataBrokerClient();
            broker.Subscribe(0, "RigctldServerEnabled", OnSettingChanged);
            broker.Subscribe(0, "RigctldServerPort", OnSettingChanged);
            broker.Subscribe(0, "ServerBindAll", OnSettingChanged);
            broker.Subscribe(1, "ConnectedRadios", OnConnectedRadiosChanged);
            broker.Subscribe(DataBroker.AllDevices, "Channels", OnChannelsChanged);
            broker.Subscribe(DataBroker.AllDevices, "Settings", OnSettingsChanged);

            int enabled = broker.GetValue<int>(0, "RigctldServerEnabled", 0);
            if (enabled == 1)
            {
                port = broker.GetValue<int>(0, "RigctldServerPort", 4532);
                bindAll = broker.GetValue<int>(0, "ServerBindAll", 0) == 1;
                Start();
            }
        }

        private void OnSettingChanged(int deviceId, string name, object data)
        {
            int enabled = broker.GetValue<int>(0, "RigctldServerEnabled", 0);
            int newPort = broker.GetValue<int>(0, "RigctldServerPort", 4532);
            bool newBindAll = broker.GetValue<int>(0, "ServerBindAll", 0) == 1;

            if (enabled == 1)
            {
                if (running && (newPort != port || newBindAll != bindAll))
                {
                    Stop();
                    port = newPort;
                    bindAll = newBindAll;
                    Start();
                }
                else if (!running)
                {
                    port = newPort;
                    bindAll = newBindAll;
                    Start();
                }
            }
            else
            {
                if (running) Stop();
            }
        }

        private void OnConnectedRadiosChanged(int deviceId, string name, object data)
        {
            activeRadioId = GetFirstConnectedRadioId();
        }

        private void OnChannelsChanged(int deviceId, string name, object data)
        {
            if (deviceId < 100) return;
            try
            {
                if (data is RadioChannelInfo[] channels && channels.Length > 0)
                {
                    if (channels[0].rx_freq > 0) cachedFrequency = channels[0].rx_freq;
                }
            }
            catch { }
        }

        private void OnSettingsChanged(int deviceId, string name, object data)
        {
            if (deviceId < 100) return;
            try
            {
                if (data is RadioSettings settings)
                {
                    if (settings.vfo1_mod_freq_x > 0) cachedFrequency = settings.vfo1_mod_freq_x;
                }
            }
            catch { }
        }

        private void Start()
        {
            if (running) return;
            try
            {
                cts = new CancellationTokenSource();
                listener = new TcpListener(bindAll ? IPAddress.Any : IPAddress.Loopback, port);
                listener.Start();
                running = true;
                serverTask = Task.Run(() => AcceptClientsAsync(cts.Token), cts.Token);
                Log($"Rigctld server started on port {port}" + (bindAll ? " (all interfaces - WARNING: no authentication, any LAN client can control PTT and frequency)" : " (loopback only)"));
            }
            catch (Exception ex)
            {
                Log($"Rigctld server start failed: {ex.Message}");
                running = false;
            }
        }

        private void Stop()
        {
            if (!running) return;
            Log("Rigctld server stopping...");
            running = false;
            SetPtt(false);
            cts?.Cancel();

            try { listener?.Stop(); } catch { }

            try { serverTask?.Wait(TimeSpan.FromSeconds(3)); }
            catch (AggregateException) { }
            catch (OperationCanceledException) { }

            foreach (var client in clients.Values)
            {
                try { client.Dispose(); } catch { }
            }
            clients.Clear();

            cts?.Dispose();
            cts = null;
            serverTask = null;
            Log("Rigctld server stopped");
        }

        private async Task AcceptClientsAsync(CancellationToken ct)
        {
            try
            {
                while (!ct.IsCancellationRequested)
                {
                    TcpClient tcpClient = await listener.AcceptTcpClientAsync();
                    if (clients.Count >= MaxClients)
                    {
                        Log("Rigctld connection rejected: max clients reached");
                        tcpClient.Close();
                        continue;
                    }
                    var handler = new RigctldClientHandler(tcpClient, this);
                    if (clients.TryAdd(handler.Id, handler))
                    {
                        Log($"Rigctld client connected: {handler.EndPoint}");
                    }
                    else
                    {
                        tcpClient.Close();
                    }
                }
            }
            catch (SocketException ex) when (ex.SocketErrorCode == SocketError.Interrupted) { }
            catch (ObjectDisposedException) { }
            catch (OperationCanceledException) { }
            catch (Exception ex)
            {
                Log($"Rigctld accept loop error: {ex.Message}");
            }
        }

        internal string ProcessCommand(string line)
        {
            if (string.IsNullOrWhiteSpace(line)) return "";

            line = line.Trim();

            // Handle extended protocol prefix
            bool extended = line.StartsWith("+");
            if (extended) line = line.Substring(1).TrimStart();

            // \dump_state — required by WSJT-X on connect
            if (line == "\\dump_state")
            {
                return GetDumpState();
            }

            // Parse command and arguments
            string cmd;
            string args = "";
            if (line.Length == 1)
            {
                cmd = line;
            }
            else if (line[0] == '\\')
            {
                // Long-form command like \set_ptt
                int spaceIdx = line.IndexOf(' ');
                if (spaceIdx > 0)
                {
                    cmd = line.Substring(0, spaceIdx);
                    args = line.Substring(spaceIdx + 1).Trim();
                }
                else
                {
                    cmd = line;
                }
            }
            else
            {
                cmd = line.Substring(0, 1);
                args = line.Length > 1 ? line.Substring(1).Trim() : "";
            }

            switch (cmd)
            {
                case "T":
                case "\\set_ptt":
                {
                    int pttVal = 0;
                    int.TryParse(args, out pttVal);
                    SetPtt(pttVal != 0);
                    return extended ? $"set_ptt: {pttVal}\nRPRT 0\n" : "RPRT 0\n"; // pttVal is int, safe from injection
                }
                case "t":
                case "\\get_ptt":
                    return extended ? $"get_ptt:\nPTT: {(pttActive ? 1 : 0)}\n" : $"{(pttActive ? 1 : 0)}\n";

                case "f":
                case "\\get_freq":
                    return extended ? $"get_freq:\nFreq: {cachedFrequency}\n" : $"{cachedFrequency}\n";

                case "F":
                case "\\set_freq":
                {
                    long freq;
                    if (long.TryParse(args, out freq) && freq > 0 && freq <= int.MaxValue)
                    {
                        cachedFrequency = freq;
                        SetRadioFrequency(freq, "A");
                    }
                    string safeArgs = args?.Replace("\r", "").Replace("\n", "").Replace("\u2028", "").Replace("\u2029", "") ?? "";
                    return extended ? $"set_freq: {safeArgs}\nRPRT 0\n" : "RPRT 0\n";
                }

                case "m":
                case "\\get_mode":
                    return extended ? "get_mode:\nMode: FM\nPassband: 15000\n" : "FM\n15000\n";

                case "M":
                case "\\set_mode":
                    { string safeArgs2 = args?.Replace("\r", "").Replace("\n", "").Replace("\u2028", "").Replace("\u2029", "") ?? ""; return extended ? $"set_mode: {safeArgs2}\nRPRT 0\n" : "RPRT 0\n"; }

                case "v":
                case "\\get_vfo":
                    return extended ? "get_vfo:\nVFO: VFOA\n" : "VFOA\n";

                case "V":
                case "\\set_vfo":
                    { string safeArgs3 = args?.Replace("\r", "").Replace("\n", "").Replace("\u2028", "").Replace("\u2029", "") ?? ""; return extended ? $"set_vfo: {safeArgs3}\nRPRT 0\n" : "RPRT 0\n"; }

                case "s":
                case "\\get_split_vfo":
                    return extended ? "get_split_vfo:\nSplit: 0\nTX VFO: VFOA\n" : "0\nVFOA\n";

                case "q":
                case "\\quit":
                    return null; // Signal disconnect

                default:
                    Log($"Rigctld unknown command: {line}");
                    return extended ? $"{cmd}:\nRPRT -1\n" : "RPRT -1\n";
            }
        }

        private string GetDumpState()
        {
            var sb = new StringBuilder();
            sb.AppendLine("2");           // Protocol version
            sb.AppendLine("2");           // Rig model (Dummy)
            sb.AppendLine("2");           // ITU region
            // Frequency range: 100kHz - 1.3GHz
            sb.AppendLine("100000.000000 1300000000.000000 0x40000000 -1 -1 0x16000003 0x3");
            sb.AppendLine("0 0 0 0 0 0 0");  // End freq range
            // TX range
            sb.AppendLine("100000.000000 1300000000.000000 0x40000000 -1 -1 0x16000003 0x3");
            sb.AppendLine("0 0 0 0 0 0 0");  // End TX range
            sb.AppendLine("0x40000000 1");    // FM mode
            sb.AppendLine("0 0");             // End mode list
            sb.AppendLine("0");               // Max RIT
            sb.AppendLine("0");               // Max XIT
            sb.AppendLine("0");               // Max IF shift
            sb.AppendLine("0");               // Announces
            sb.AppendLine("0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0"); // Preamp
            sb.AppendLine("0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0"); // Attenuator
            sb.AppendLine("0x1e000000");      // has_get_func
            sb.AppendLine("0x1e000000");      // has_set_func
            sb.AppendLine("0x40000000");      // has_get_level
            sb.AppendLine("0x40000000");      // has_set_level
            sb.AppendLine("0");               // has_get_parm
            sb.AppendLine("0");               // has_set_parm
            sb.AppendLine("vfo_op=0x0");
            sb.AppendLine("done");
            return sb.ToString();
        }

        private void SetRadioFrequency(long freqHz, string vfo)
        {
            // Validate frequency fits in int (max ~2.1 GHz) to prevent integer overflow
            if (freqHz <= 0 || freqHz > int.MaxValue) return;

            int radioId = activeRadioId;
            if (radioId < 0) radioId = GetFirstConnectedRadioId();
            if (radioId < 0) return;

            var info = broker.GetValue<RadioDevInfo>(radioId, "Info", null);
            if (info == null) return;

            int scratchIndex = info.channel_count - 1;
            var scratch = new RadioChannelInfo();
            scratch.channel_id = scratchIndex;
            scratch.rx_freq = (int)freqHz;
            scratch.tx_freq = (int)freqHz;
            scratch.rx_mod = Radio.RadioModulationType.FM;
            scratch.tx_mod = Radio.RadioModulationType.FM;
            scratch.bandwidth = Radio.RadioBandwidthType.WIDE;
            scratch.name_str = "QF";

            broker.Dispatch(radioId, "WriteChannel", scratch, store: false);
            string eventName = (vfo == "B") ? "ChannelChangeVfoB" : "ChannelChangeVfoA";
            broker.Dispatch(radioId, eventName, scratchIndex, store: false);
            Log($"Rigctld set_freq: {freqHz} Hz → scratch channel {scratchIndex} on VFO {vfo}");
        }

        internal void SetPtt(bool on)
        {
            lock (pttLock)
            {
                bool wasActive = pttActive;
                pttActive = on;

                if (on && !wasActive)
                {
                    // Start dispatching silence to keep radio keyed
                    pttSilenceTimer = new Timer(DispatchSilence, null, 0, 80);
                    pttTimeoutTimer = new Timer(PttTimeoutCallback, null, PttTimeoutMs, Timeout.Infinite);
                    Log("Rigctld PTT ON");
                    broker?.Dispatch(1, "ExternalPttState", true, store: false);
                }
                else if (!on && wasActive)
                {
                    pttSilenceTimer?.Dispose();
                    pttSilenceTimer = null;
                    pttTimeoutTimer?.Dispose();
                    pttTimeoutTimer = null;
                    Log("Rigctld PTT OFF");
                    broker?.Dispatch(1, "ExternalPttState", false, store: false);
                }
            }
        }

        private void PttTimeoutCallback(object state)
        {
            Log("Rigctld PTT auto-released after timeout");
            SetPtt(false);
        }

        private void DispatchSilence(object state)
        {
            if (!pttActive) return;
            int radioId = activeRadioId;
            if (radioId < 0) radioId = GetFirstConnectedRadioId();
            if (radioId < 0) return;

            // 100ms of 32kHz 16-bit mono silence = 6400 bytes
            byte[] silence = new byte[6400];
            broker?.Dispatch(radioId, "TransmitVoicePCM", silence, store: false);
        }

        private int GetFirstConnectedRadioId()
        {
            var radios = broker.GetValue<object>(1, "ConnectedRadios", null);
            if (radios is System.Collections.IEnumerable enumerable)
            {
                foreach (var item in enumerable)
                {
                    if (item == null) continue;
                    var prop = item.GetType().GetProperty("DeviceId");
                    if (prop != null)
                    {
                        object val = prop.GetValue(item);
                        if (val is int id && id > 0) return id;
                    }
                }
            }
            return -1;
        }

        internal void RemoveClient(Guid clientId)
        {
            if (clients.TryRemove(clientId, out _))
            {
                Log($"Rigctld client disconnected: {clientId}");
                // Release PTT if no clients remain (prevents stuck transmit on disconnect)
                if (pttActive && clients.IsEmpty)
                {
                    Log("Rigctld: releasing PTT (last client disconnected)");
                    SetPtt(false);
                }
            }
        }

        private void Log(string message)
        {
            broker?.LogInfo(message);
        }

        public void Dispose()
        {
            Stop();
            broker?.Dispose();
        }
    }

    internal class RigctldClientHandler : IDisposable
    {
        private readonly TcpClient client;
        private readonly NetworkStream stream;
        private readonly RigctldServer server;
        private readonly CancellationTokenSource cts = new CancellationTokenSource();

        public Guid Id { get; }
        public IPEndPoint EndPoint => (IPEndPoint)client.Client.RemoteEndPoint;

        public RigctldClientHandler(TcpClient client, RigctldServer server)
        {
            Id = Guid.NewGuid();
            this.client = client;
            this.stream = client.GetStream();
            this.server = server;

            Task.Run(ReceiveLoopAsync, cts.Token);
        }

        private async Task ReceiveLoopAsync()
        {
            var writer = new StreamWriter(stream, Encoding.ASCII) { AutoFlush = true };

            try
            {
                // Use a length-limited line reader to prevent unbounded memory allocation
                // from oversized lines (ReadLineAsync buffers the entire line first)
                while (!cts.Token.IsCancellationRequested)
                {
                    string line = await ReadLineLimitedAsync(stream, 1024, cts.Token);
                    if (line == null) break; // Disconnected or oversized

                    string response = server.ProcessCommand(line);
                    if (response == null) break; // quit command

                    if (response.Length > 0)
                    {
                        await writer.WriteAsync(response);
                    }
                }
            }
            catch (OperationCanceledException) { }
            catch (IOException) { }
            catch (Exception) { }

            Disconnect();
        }

        /// <summary>
        /// Reads a line from the stream with a maximum byte limit to prevent unbounded memory allocation.
        /// Returns null on disconnect or if the line exceeds maxLength.
        /// </summary>
        private static async Task<string> ReadLineLimitedAsync(System.Net.Sockets.NetworkStream stream, int maxLength, CancellationToken ct)
        {
            var sb = new System.Text.StringBuilder(128);
            byte[] buf = new byte[1];
            // Idle timeout: disconnect clients that stall without sending data
            using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            timeoutCts.CancelAfter(30000); // 30 second idle timeout
            while (!timeoutCts.Token.IsCancellationRequested)
            {
                int read = await stream.ReadAsync(buf, 0, 1, timeoutCts.Token);
                if (read == 0) return sb.Length > 0 ? sb.ToString() : null;
                char c = (char)buf[0];
                if (c == '\n') return sb.ToString().TrimEnd('\r');
                sb.Append(c);
                if (sb.Length > maxLength) return null; // Oversized — disconnect
            }
            return null;
        }

        private void Disconnect()
        {
            if (!cts.IsCancellationRequested) cts.Cancel();
            server.RemoveClient(Id);
        }

        public void Dispose()
        {
            Disconnect();
            stream?.Dispose();
            client?.Dispose();
            cts?.Dispose();
        }
    }
}

File: htcommander_flutter/lib/servers/rigctld_server.dart
/*
Copyright 2026 Ylian Saint-Hilaire
Licensed under the Apache License, Version 2.0 (the "License");
http://www.apache.org/licenses/LICENSE-2.0
*/

import 'dart:async';
import 'dart:convert';
import 'dart:io';

import '../core/data_broker.dart';
import '../core/data_broker_client.dart';

/// Hamlib rigctld TCP text protocol server.
///
/// Used by fldigi, WSJT-X, Direwolf, VaraFM.
/// Port of HTCommander.Core/Utils/RigctldServer.cs.
class RigctldServer {
  final DataBrokerClient _broker = DataBrokerClient();
  ServerSocket? _serverSocket;
  final Map<String, _RigctldClientHandler> _clients = {};
  static const int _maxClients = 10;
  int _port = 4532;
  bool _running = false;
  bool _bindAll = false;
  bool _pttActive = false;
  Timer? _pttSilenceTimer;
  Timer? _pttTimeoutTimer;
  static const int _pttTimeoutMs = 30000;
  int _cachedFrequency = 145500000;
  int _activeRadioId = -1;

  bool get pttActive => _pttActive;
  int get cachedFrequency => _cachedFrequency;

  RigctldServer() {
    _broker.subscribe(0, 'RigctldServerEnabled', _onSettingChanged);
    _broker.subscribe(0, 'RigctldServerPort', _onSettingChanged);
    _broker.subscribe(0, 'ServerBindAll', _onSettingChanged);
    _broker.subscribe(1, 'ConnectedRadios', _onConnectedRadiosChanged);
    _broker.subscribe(DataBroker.allDevices, 'Channels', _onChannelsChanged);
    _broker.subscribe(DataBroker.allDevices, 'Settings', _onSettingsChanged);

    final enabled = _broker.getValue<int>(0, 'RigctldServerEnabled', 0);
    if (enabled == 1) {
      _port = _broker.getValue<int>(0, 'RigctldServerPort', 4532);
      _bindAll = _broker.getValue<int>(0, 'ServerBindAll', 0) == 1;
      _start();
    }
  }

  void _onSettingChanged(int deviceId, String name, Object? data) {
    final enabled = _broker.getValue<int>(0, 'RigctldServerEnabled', 0);
    final newPort = _broker.getValue<int>(0, 'RigctldServerPort', 4532);
    final newBindAll = _broker.getValue<int>(0, 'ServerBindAll', 0) == 1;

    if (enabled == 1) {
      if (_running && (newPort != _port || newBindAll != _bindAll)) {
        _stop();
        _port = newPort;
        _bindAll = newBindAll;
        _start();
      } else if (!_running) {
        _port = newPort;
        _bindAll = newBindAll;
        _start();
      }
    } else {
      if (_running) _stop();
    }
  }

  void _onConnectedRadiosChanged(int deviceId, String name, Object? data) {
    _activeRadioId = _getFirstConnectedRadioId();
  }

  void _onChannelsChanged(int deviceId, String name, Object? data) {
    if (deviceId < 100) return;
    if (data is List && data.isNotEmpty) {
      final first = data[0];
      if (first is Map && (first['rx_freq'] as int? ?? 0) > 0) {
        _cachedFrequency = first['rx_freq'] as int;
      }
    }
  }

  void _onSettingsChanged(int deviceId, String name, Object? data) {
    if (deviceId < 100) return;
    if (data is Map) {
      final freq = data['vfo1_mod_freq_x'];
      if (freq is int && freq > 0) _cachedFrequency = freq;
    }
  }

  Future<void> _start() async {
    if (_running) return;
    try {
      final address =
          _bindAll ? InternetAddress.anyIPv4 : InternetAddress.loopbackIPv4;
      _serverSocket = await ServerSocket.bind(address, _port);
      _running = true;
      _log('Rigctld server started on port $_port'
          '${_bindAll ? " (all interfaces)" : " (loopback only)"}');

      _serverSocket!.listen(
        (socket) {
          if (_clients.length >= _maxClients) {
            _log('Rigctld connection rejected: max clients reached');
            socket.close();
            return;
          }
          final handler = _RigctldClientHandler(socket, this);
          _clients[handler.id] = handler;
          _log('Rigctld client connected');
        },
        onError: (_) {},
        cancelOnError: false,
      );
    } catch (e) {
      _log('Rigctld server start failed: $e');
      _running = false;
    }
  }

  void _stop() {
    if (!_running) return;
    _log('Rigctld server stopping...');
    _running = false;
    setPtt(false);
    _serverSocket?.close();
    _serverSocket = null;

    for (final client in _clients.values) {
      client.dispose();
    }
    _clients.clear();
    _log('Rigctld server stopped');
  }

  String processCommand(String line) {
    if (line.trim().isEmpty) return '';

    var cmd = line.trim();
    bool extended = cmd.startsWith('+');
    if (extended) cmd = cmd.substring(1).trimLeft();

    // dump_state
    if (cmd == r'\dump_state') return _getDumpState();

    String command;
    String args = '';
    if (cmd.length == 1) {
      command = cmd;
    } else if (cmd.startsWith(r'\')) {
      final spaceIdx = cmd.indexOf(' ');
      if (spaceIdx > 0) {
        command = cmd.substring(0, spaceIdx);
        args = cmd.substring(spaceIdx + 1).trim();
      } else {
        command = cmd;
      }
    } else {
      command = cmd.substring(0, 1);
      args = cmd.length > 1 ? cmd.substring(1).trim() : '';
    }

    switch (command) {
      case 'T':
      case r'\set_ptt':
        final pttVal = int.tryParse(args) ?? 0;
        setPtt(pttVal != 0);
        return extended ? 'set_ptt: $pttVal\nRPRT 0\n' : 'RPRT 0\n';
      case 't':
      case r'\get_ptt':
        final v = _pttActive ? 1 : 0;
        return extended ? 'get_ptt:\nPTT: $v\n' : '$v\n';
      case 'f':
      case r'\get_freq':
        return extended
            ? 'get_freq:\nFreq: $_cachedFrequency\n'
            : '$_cachedFrequency\n';
      case 'F':
      case r'\set_freq':
        final freq = int.tryParse(args);
        if (freq != null && freq > 0 && freq <= 2147483647) {
          _cachedFrequency = freq;
          _setRadioFrequency(freq, 'A');
        }
        final safeArgs = _sanitize(args);
        return extended
            ? 'set_freq: $safeArgs\nRPRT 0\n'
            : 'RPRT 0\n';
      case 'm':
      case r'\get_mode':
        return extended
            ? 'get_mode:\nMode: FM\nPassband: 15000\n'
            : 'FM\n15000\n';
      case 'M':
      case r'\set_mode':
        final safeArgs = _sanitize(args);
        return extended ? 'set_mode: $safeArgs\nRPRT 0\n' : 'RPRT 0\n';
      case 'v':
      case r'\get_vfo':
        return extended ? 'get_vfo:\nVFO: VFOA\n' : 'VFOA\n';
      case 'V':
      case r'\set_vfo':
        final safeArgs = _sanitize(args);
        return extended ? 'set_vfo: $safeArgs\nRPRT 0\n' : 'RPRT 0\n';
      case 's':
      case r'\get_split_vfo':
        return extended
            ? 'get_split_vfo:\nSplit: 0\nTX VFO: VFOA\n'
            : '0\nVFOA\n';
      case 'q':
      case r'\quit':
        return ''; // Signal disconnect handled by caller
      default:
        return extended ? '$command:\nRPRT -1\n' : 'RPRT -1\n';
    }
  }

  String _getDumpState() {
    final sb = StringBuffer();
    sb.writeln('2'); // Protocol version
    sb.writeln('2'); // Rig model
    sb.writeln('2'); // ITU region
    sb.writeln(
        '100000.000000 1300000000.000000 0x40000000 -1 -1 0x16000003 0x3');
    sb.writeln('0 0 0 0 0 0 0');
    sb.writeln(
        '100000.000000 1300000000.000000 0x40000000 -1 -1 0x16000003 0x3');
    sb.writeln('0 0 0 0 0 0 0');
    sb.writeln('0x40000000 1');
    sb.writeln('0 0');
    sb.writeln('0');
    sb.writeln('0');
    sb.writeln('0');
    sb.writeln('0');
    sb.writeln(
        '0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0');
    sb.writeln(
        '0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0');
    sb.writeln('0x1e000000');
    sb.writeln('0x1e000000');
    sb.writeln('0x40000000');
    sb.writeln('0x40000000');
    sb.writeln('0');
    sb.writeln('0');
    sb.writeln('vfo_op=0x0');
    sb.writeln('done');
    return sb.toString();
  }

  void _setRadioFrequency(int freqHz, String vfo) {
    if (freqHz <= 0 || freqHz > 2147483647) return;

    int radioId = _activeRadioId;
    if (radioId < 0) radioId = _getFirstConnectedRadioId();
    if (radioId < 0) return;

    final info = _broker.getValueDynamic(radioId, 'Info');
    if (info == null) return;

    final channelCount =
        (info is Map ? info['channel_count'] as int? : null) ?? 0;
    if (channelCount <= 0) return;

    final scratchIndex = channelCount - 1;
    _broker.dispatch(
        radioId,
        'WriteChannel',
        {
          'channel_id': scratchIndex,
          'rx_freq': freqHz,
          'tx_freq': freqHz,
          'name_str': 'QF',
        },
        store: false);

    final eventName =
        vfo == 'B' ? 'ChannelChangeVfoB' : 'ChannelChangeVfoA';
    _broker.dispatch(radioId, eventName, scratchIndex, store: false);
    _log('Rigctld set_freq: $freqHz Hz on VFO $vfo');
  }

  void setPtt(bool on) {
    final wasActive = _pttActive;
    _pttActive = on;

    if (on && !wasActive) {
      _pttSilenceTimer?.cancel();
      _pttSilenceTimer =
          Timer.periodic(const Duration(milliseconds: 80), (_) {
        _dispatchSilence();
      });
      _pttTimeoutTimer?.cancel();
      _pttTimeoutTimer = Timer(
          const Duration(milliseconds: _pttTimeoutMs), _pttTimeoutCallback);
      _log('Rigctld PTT ON');
      _broker.dispatch(1, 'ExternalPttState', true, store: false);
    } else if (!on && wasActive) {
      _pttSilenceTimer?.cancel();
      _pttSilenceTimer = null;
      _pttTimeoutTimer?.cancel();
      _pttTimeoutTimer = null;
      _log('Rigctld PTT OFF');
      _broker.dispatch(1, 'ExternalPttState', false, store: false);
    }
  }

  void _pttTimeoutCallback() {
    _log('Rigctld PTT auto-released after timeout');
    setPtt(false);
  }

  void _dispatchSilence() {
    if (!_pttActive) return;
    int radioId = _activeRadioId;
    if (radioId < 0) radioId = _getFirstConnectedRadioId();
    if (radioId < 0) return;

    // 100ms of 32kHz 16-bit mono silence = 6400 bytes
    final silence = List<int>.filled(6400, 0);
    _broker.dispatch(radioId, 'TransmitVoicePCM', silence, store: false);
  }

  int _getFirstConnectedRadioId() {
    final radios = _broker.getValueDynamic(1, 'ConnectedRadios');
    if (radios is List) {
      for (final item in radios) {
        if (item is Map) {
          final id = item['deviceId'] as int?;
          if (id != null && id > 0) return id;
        }
      }
    }
    return -1;
  }

  void removeClient(String clientId) {
    if (_clients.remove(clientId) != null) {
      _log('Rigctld client disconnected');
      if (_pttActive && _clients.isEmpty) {
        _log('Rigctld: releasing PTT (last client disconnected)');
        setPtt(false);
      }
    }
  }

  String _sanitize(String s) {
    return s
        .replaceAll('\r', '')
        .replaceAll('\n', '')
        .replaceAll('\u2028', '')
        .replaceAll('\u2029', '');
  }

  void _log(String message) {
    _broker.logInfo(message);
  }

  void dispose() {
    _stop();
    _broker.dispose();
  }
}

class _RigctldClientHandler {
  final Socket _socket;
  final RigctldServer _server;
  final String id;
  final StringBuffer _lineBuffer = StringBuffer();
  Timer? _idleTimer;
  static const int _maxLineLength = 1024;
  static const Duration _idleTimeout = Duration(seconds: 30);
  bool _disposed = false;

  _RigctldClientHandler(this._socket, this._server)
      : id = DateTime.now().microsecondsSinceEpoch.toString() {
    _resetIdleTimer();
    _socket.listen(
      _onData,
      onError: (_) => _disconnect(),
      onDone: _disconnect,
      cancelOnError: false,
    );
  }

  void _onData(List<int> data) {
    _resetIdleTimer();
    final chars = ascii.decode(data, allowInvalid: true);
    _lineBuffer.write(chars);

    while (true) {
      final buf = _lineBuffer.toString();
      final nlIdx = buf.indexOf('\n');
      if (nlIdx < 0) break;

      final line = buf.substring(0, nlIdx).trimRight();
      _lineBuffer.clear();
      if (nlIdx + 1 < buf.length) {
        _lineBuffer.write(buf.substring(nlIdx + 1));
      }

      if (_lineBuffer.length > _maxLineLength) {
        _disconnect();
        return;
      }

      final response = _server.processCommand(line);
      if (response.isEmpty && (line.trim() == 'q' || line.trim() == r'\quit')) {
        _disconnect();
        return;
      }
      if (response.isNotEmpty) {
        try {
          _socket.write(response);
        } catch (_) {
          _disconnect();
          return;
        }
      }
    }

    if (_lineBuffer.length > _maxLineLength) {
      _disconnect();
    }
  }

  void _resetIdleTimer() {
    _idleTimer?.cancel();
    _idleTimer = Timer(_idleTimeout, _disconnect);
  }

  void _disconnect() {
    if (_disposed) return;
    _disposed = true;
    _idleTimer?.cancel();
    try {
      _socket.close();
    } catch (_) {}
    _server.removeClient(id);
  }

  void dispose() {
    _disconnect();
  }
}
